src/epics/revisions.js

```diff
@@ -1,68 +1,254 @@
+};
+
+// Dispatch an action when the query changes from ready to not ready (or vice-versa)
+export const revisionsReady = ( action$, store ) => (
 	action$
-		.filter( ( action ) => [ 'QUERY_UPDATE', 'QUERY_SET_VALUE', 'WIKIS_SET' ].includes( action.type ) )
-		.filter( () => !!store.getState().query.wiki && store.getState().wikis.size > 0 && store.getState().query.user.size > 0 )
-		.flatMap( () => Observable.of( RevisionsActions.fetchRevisions() ) )
+		.filter( ( action ) => [ 'QUERY_WIKI_CHANGE', 'QUERY_USER_CHANGE' ].includes( action.type ) )
+		// Determine if the query is ready or not.
+		.map( () => !!store.getState().query.wiki && !store.getState().wikis.isEmpty() && !store.getState().query.user.isEmpty() )
+		// Wait until the status has changed.
+		.filter( ready => !( ready && store.getState().revisions.status === 'ready' ) )
+		.filter( ready => !( !ready && store.getState().revisions.status === 'notready' ) )
+		.map( ( ready ) => {
+			return ready ? RevisionsActions.setStatusReady() : RevisionsActions.setStatusNotReady();
+		} )
 );
 
-export const hideLoader = ( action$, store ) => (
-	action$
-		.filter( ( action ) => [ 'QUERY_UPDATE', 'QUERY_SET_VALUE', 'WIKIS_SET' ].includes( action.type ) )
-		.filter( () => !store.getState().query.wiki || !store.getState().query.user.size )
-		.flatMap( () => Observable.of( RevisionsActions.setStatusReady() ) )
+export const shouldFetchRevisions = ( action$, store ) => (
+	action$.filter( ( action ) => [
+		'QUERY_WIKI_CHANGE',
+		'QUERY_USER_CHANGE',
+		'QUERY_START_DATE_CHANGE',
+		'QUERY_END_DATE_CHANGE',
+		'REVISIONS_READY'
+	].includes( action.type ) )
+		.filter( () => store.getState().revisions.status !== 'notready' )
+		.flatMap( () => {
+			const users = store.getState().query.user
+				.filter( user => !store.getState().revisions.cont.keySeq().includes( user ) );
+
+			return Observable.of( users );
+		} )
+		.filter( users => !!users.size )
+		.flatMap( ( users ) => Observable.of( RevisionsActions.fetchRevisions( users ) ) )
+);
+
+export const revisionStatus = ( action$, store ) => (
+	action$.ofType( 'REVISIONS_ADD' )
+		.map( ( action ) => {
+			const cont = store.getState().revisions.cont;
+
+			if ( cont.filter( c => c !== false ).isEmpty() ) {
+				return RevisionsActions.setStatusDone();
+			}
+
+			if ( action.revisions.isEmpty() ) {
+				return RevisionsActions.setStatusReady();
+			}
+
+			const users = store.getState().query.user;
+			const revisions = store.getState().revisions.list;
+			const pages = store.getState().pages;
+			const last = getLastRevision( revisions, users, cont );
+
+			if ( !last ) {
+				return RevisionsActions.setStatusReady();
+			}
+
+			if ( !getRevisions( users, revisions, last, pages ).isEmpty() ) {
+				return RevisionsActions.setStatusReady();
+			}
+
+			return RevisionsActions.fetchRevisions( new Set( [ last.user ] ) );
+		} )
 );
 
 export const fetchRevisions = ( action$, store ) => (
```

Comment:
This became massive. If possible it would be a good idea to split it up in different functions and make more easy to read

Reply:
@dayllanmaza Do you mind if this gets fixed in T180083 ? I have several changes in there too.

Reply:
I don't